### ISSUE TITLE
Track activity on items

### ISSUE THREAD

**Issue by xmatthewx** [mozilla/network-pulse#251]
We should gather data so we can display things like Fav count, popular items, etc. 

See: https://github.com/mozilla/network-pulse/issues/236#issuecomment-267435069

cc @alanmoo

**Comment by alanmoo** [comment]
A thought about Google Analytics- I'm not convinced it's accurate given our specific target audience. Perhaps we should investigate a self-hosted analytics solution at some point? It seems like this might be a good project to test something on (we can do that after launch though).

**Comment by xmatthewx** [comment]
Though we'll use analytics for internal learning, this is specific about us storing info on activity within our DB. When a user Favs an item, for example, we'll want to store that item ID on their account AND we can count aggregate info on items themselves. How many people Fav'd it? This would be baked into the API. I don't think we'll want or need a 3rd party tool to display ♡15 on an item.

A _trending_ rather than _latest_ view could sort by some math on views + share intent + favs + visit url, weighted toward recent activity. 

We're obviously not ready to implement a trending view now. But it would valuable to gather numbers from the beginning.

**Comment by alanmoo** [comment]
I'm not quite sure how this data would be accurate if users of the site aren't necessarily logged in. We could keep track of the number of times a button's been clicked, but that says nothing of people changing browsers, emptying cache, etc.

**Comment by Pomax** [comment]
I would strongly advocate tying any kind of likes/faves/etc behind a login, similar to hackernews, reddit, imgur, and other popular "user submission" sites. You can enjoy the content, but if you want to build the community (by 'rewarding' submitted with likes/faves/etc) you should be an identifiable party.

**Comment by xmatthewx** [comment]
Yes, very good points. 

While we build an audience in 2017, anonymous views can help us build curation to build an audience to build more meaningful activity. 🐔 🍳 

Maybe we can do a little of this, then turn it off when the bots arrive? A threshold of minimum favs by known users could help filter out views from people trying to game the system.

**Comment by Pomax** [comment]
Oh that's a good point - we probably want to track view count in a way that lets us see how many show up "in searches" as well as how many are clicked-through. On community-voted sites the main page tends to be the highest voted content (based on some magic internal score), but since we won't have a "community" as such at first, pure "how many times has this thing been on people's screens" will be an interesting statistic to look at (although mostly as a rough indicator, due to page reloads, bot scrapes, etc)

**Comment by alanmoo** [comment]
I'm working on a PR that adds tracking to the views themselves (such as `/entries/` or `/entries/{#}`). It works without much effort, though I'm not sure it quite covers everything we want to do. Along with the associated metadata, it also captures the server response which we could search through later on. However, it seems less-than-trivial to add some sort of indicator in the db to indicate which entries were viewed as part of that list (though we can see which ones were specifically requested in a detail view). Once we implement some sort of server side favorites, it will also be straightforward to track them.

**Comment by alanmoo** [comment]
This appears to be the last item that's blocking us from completing the back end. I don't believe we've got enough of a plan to really act on it, when considering the comments @Pomax made in [this PR](https://github.com/mozilla/network-pulse-api/pull/8)

**Comment by alanmoo** [comment]
Or maybe @cadecairos you do?

**Comment by alanmoo** [comment]
Alright I've come up with some thoughts:

Fav count for authed users: simple many to many relationship from user account to "liked" entries. We can display that as a count on the entries or a list for the user.

"Shown in group" counter on the entry model, incremented every time an API call shows this entry in a list of results (i.e.`GET /entries/`)

"Viewed individually" counter on each entry, incremented any time a specific entry is fetched (`GET /entries/2` for example)

And maybe an "anonymous favs" counter on each entry, accessed by `/entries/2/fav` and `/entries/2/unfav`?

@xmatthewx the only thing I'm not sure about is your mention of "share intent" above. Can you look through this list and let me know if it sounds about right and I'll start working on it? Also open to feedback from others re: how to best implement this, I'm kind of flying blind and just guessing. @gideonthomas maybe you can comment?

**Comment by xmatthewx** [comment]
> This appears to be the last item that's blocking us from completing the back end.

Really? That's awesome news. 

This list looks great @alanmoo. Let's discuss whether to push this as a follow on item. Might be worth wrapping up the backend so we can start testing everything. Unless this will be fairly quick and easy, stats could wait until early Feb. Curious what you and @Pomax think. 

"Share intent" would count clicks on Link (implemented) or Facebook/Twitter (not yet implemented). This could be authed only, or authed and anonymous separately. We won't know if the share actually happened, but this should be a good proxy for spotting trends.

**Comment by cadecairos** [comment]
I'm not entirely sold on building in an analytics system beyond simple upvotes[1], when we can use GA to see what's popular and manually promote popular content via the django admin interface..

I think the effort required and complexity of rolling custom analytics directly into the api outweighs using what's readily available from our GA account.

[1] should be easy, especially if auth'd

**Comment by xmatthewx** [comment]
Good feedback. We can work with our Featured view and later build a manual Trending view later until we achieve a need for something more sophisticated. 

Before we close this.... 
 1. Let's make sure our GA implementation is tight. On Webmaker app, it was difficult to see trends. Project views were buried in events without project titles. GA on the mozfest event app was better. Let's make this the best. Can we have Views as Pageviews in GA? 
 2. Let's look at the list and identify anything we couldn't do with GA (e.g. display a fav count)

Fav count – We aren't going to display that count right now. If we can generate a fav-per-project count later from each user's fav list, then we can table this until we decide we want it.

**Comment by cadecairos** [comment]
> Fav count – We aren't going to display that count right now. If we can generate a fav-per-project count later from each user's fav list, then we can table this until we decide we want it.

as long as the fav is just a simple table row showing userID and Project ID, we can extract whatever we need via query (favs for a user, favs for a project, etc.)

**Comment by alanmoo** [comment]
Yeah, I'll start with the authenticated user favorites; that seems like the most logical place to begin building out any sort of useful data.

**Comment by Pomax** [comment]
+1 - though we'll want a triplet `{user id, entry id, time of faving}` rather than just the user and entry tuple, so that we can trivially grep for popularity-over-time instead of that being literally impossible =)

**Comment by Pomax** [comment]
As per the current pulse API rewrite (and pending PRs at the time of writing this comment) we have access to the following information in the database:

- individual bookmarks placed by users
- as direct derivative: number of bookmarks by users on an entry
- as direct derivative: for individual entries, whether it is bookmarked by a viewing user or not
- creation timestamps for each user/entry bookmark relation

Given this information, we can set up queries that let us determine:

- entry rankings based on bookmarks
- most bookmarked entries for specific days
- trending entries (at least trending in terms of bookmarking)
- user bookmarking behaviour measured against age of an entry
- a ton of other things that temporally link entries, users, and bookmarks =)

Queries that would require additional data logging:

- determine what percentage of entries requested by a user are "fresh" vs. "previously bookmarked" (might require an unacceptable kind of logging)

The API server does not/cannot track:

- bookmarks by anonymous users (as a purely client-side data point, we can't track that in the database. At least not until the anonymous user logs into the app, at which point we'd sync the bookmarks - raises a possible issue around syncing "with or without" correct timestamps)

**Comment by xmatthewx** [comment]
Thanks. This is an awesome review. 

I'll file a ticket for the backlog to think about an MVP dashboard for when we've got some data to surface.

**Comment by alanmoo** [comment]
Can we close this now then?

**Comment by xmatthewx** [comment]
I think yes. Is there other things we're tracking in the DB? Or is everything else in analytics? (e.g. item views)

**Comment by Pomax** [comment]
For reliability I would says item views etc should be client side, as the client is able to tell whether a user is *actually* looking at items etc. whereas the API server can only know that "something" requested a set of entries, without any knowledge of whether that set is going to be further filtered down or even viewed at all.

**Comment by xmatthewx** [comment]
Thanks for the discussion. I think this is good and done. I filed a new ticket specific to anything lingering for google analytics.
